### ISSUE TITLE
grad returns non-deterministic values for some multi-dimentional inputs

### ISSUE THREAD

**Issue by grwlf** [PennyLaneAI/catalyst#111]
Consider the following program:

``` python
import jax.numpy as jnp
from catalyst import qjit, grad as grad

def func(p1):
  return jnp.stack([1*p1, 2*p1, 3*p1])

a = jnp.zeros([2], dtype=float)
r = qjit(grad(func, argnum=[0]))(a)
print(type(r), r.shape)
print(r)
```

While not clearly documented, we expected `r` to be a deterministic tensor representing the func's jacobian. In fact, it contains some non-deterministic values:

```
<class 'jaxlib.xla_extension.Array'> (2, 3, 2)
[[[1.00000000e+000 1.00000000e+000]
  [0.00000000e+000 0.00000000e+000]
  [4.66239210e-303 4.66239210e-303]]

 [[0.00000000e+000 0.00000000e+000]
  [2.00000000e+000 2.00000000e+000]
  [4.66239175e-303 4.66239175e-303]]]
```
where the imprecise numbers change from run to run, and sometimes contain nans (uninitialized?).

Note, that `jax.jit(jax.jacfwd(func, argnums=[0]))(a)` returns a deterministic tensor of shape `(3,2,2)`.
The catalyst is bd8eb1149c

**Comment by grwlf** [comment]
Here is some additional information on this issue:
1. Disabling the bufferization inlining (as suggested by @erick-xanadu) makes the compiler output correct results _in most cases_. We put a Catalyst version with the inlining disabled as d934fa70e09 .  
2. For this modified Catalyst, some tests always seem to be correct, but others are sometimes wrong.
3. An example of the flickering test is shown below. The fact that `f` returns multiple outputs seems to be important. We didn't see incorrect results from single-output versions of `f`. 

``` python
import jax.numpy as jnp
from catalyst import qjit, grad as C_grad
from numpy.testing import assert_allclose

def f(x):
    return (x, jnp.stack([1*x, 2*x, 3*x]))

@qjit
def C_workflow():
    return C_grad(f, method='fd', argnum=[0])(jnp.zeros([4], dtype=float))

# Check the consistency
r0 = C_workflow()
for i in range(10):
    print(f"Iteration {i}")
    r1 = C_workflow()

    for a,b in zip(r0,r1):
        assert_allclose(a, b, rtol=1e-6, atol=1e-6)

```
An example output:
``` result
Iteration 0
Iteration 1
Iteration 2
Traceback (most recent call last):
  File "/workspace/src/vjp_study/issue1.py", line 19, in <module>
    assert_allclose(a, b, rtol=1e-6, atol=1e-6)
  File "/opt/venv/lib/python3.10/site-packages/numpy/testing/_private/utils.py", line 1527, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/opt/venv/lib/python3.10/site-packages/numpy/testing/_private/utils.py", line 844, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-06, atol=1e-06

Mismatched elements: 2 / 16 (12.5%)
Max absolute difference: 1.
Max relative difference: 2.13755206e+302
 x: array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])
 y: array([[4.678249e-303, 0.000000e+000, 0.000000e+000, 0.000000e+000],
       [4.678249e-303, 0.000000e+000, 0.000000e+000, 0.000000e+000],
       [4.678249e-303, 0.000000e+000, 1.000000e+000, 0.000000e+000],
       [4.678249e-303, 0.000000e+000, 0.000000e+000, 1.000000e+000]])


```

**Comment by erick-xanadu** [comment]
@grwlf can you show here the LLVM-IR?

**Comment by dime10** [comment]
Resolved by #121
